sql: improve partial stats description

veronicavaldez opened the pull request:
Previously, the job description for AUTO CREATE PARTIAL STATS defaulted to a raw SQL description, making it harder to read and difficult to distinguish these jobs in system.jobs.

This was inadequate because users need clear job descriptions to differentiate between full and partial automatic statistics updates, especially for debugging or monitoring job execution.

To address this, this patch updates the job description to include plain English and explicitly indicate that the job is for automatic partial statistics collection, improving clarity in system.jobs.

Fixes https://github.com/cockroachdb/cockroach/issues/140054.

Release note (sql change): The job description for AUTO CREATE PARTIAL STATS now clearly indicates its purpose in plain English, improving system.jobs visibility and debugging.

veronicavaldez committed:
sql: improve auto partial stats job description

Previously, the job description for AUTO CREATE PARTIAL STATS defaulted to a
raw SQL description, making it harder to read and difficult to distinguish
these jobs in system.jobs.

This was inadequate because users need clear job descriptions to differentiate
between full and partial automatic statistics updates, especially for debugging
or monitoring job execution.

To address this, this patch updates the job description to include plain English
and explicitly indicate that the job is for automatic partial statistics
collection, improving clarity in system.jobs.

Fixes #140054.

Release note (sql change): The job description for AUTO CREATE PARTIAL STATS
now clearly indicates its purpose in plain English, improving system.jobs
visibility and debugging.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/143283)
<!-- Reviewable:end -->

rytaft reviewed (commented):
Thanks for the contribution, @veronicavaldez! This is definitely on the right track, but there are a few issues with the test that I've called out below. Please also squash your commits.

> Reviewed 2 of 2 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/143283#-:-OM7nxiN6dFhCpr3XSlM:b-b0eml0)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @veronicavaldez)

___
*[`pkg/sql/create_stats_test.go` line 247 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/143283#-OM7mUYI3nG0AP2GB17C:-OM7mUYI3nG0AP2GB17D:b64roy) ([raw file](https://github.com/cockroachdb/cockroach/blob/a258eb30f327ab522c5950f2dab6f5565683a373/pkg/sql/create_stats_test.go#L247)):*
> ```Go
> 	sqlRunner := sqlutils.MakeSQLRunner(sqlDB)
> 
> 	// Enable both automatic statistics collection and partial stats collection
> ```

nit: end comments with a period (here and below).
___
*[`pkg/sql/create_stats_test.go` line 249 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/143283#-OM7m9BT5KEFrkQFuLsm:-OM7m9BT5KEFrkQFuLsn:b-q4u0dj) ([raw file](https://github.com/cockroachdb/cockroach/blob/a258eb30f327ab522c5950f2dab6f5565683a373/pkg/sql/create_stats_test.go#L249)):*
> ```Go
> 	// Enable both automatic statistics collection and partial stats collection
> 	sqlRunner.Exec(t, `SET CLUSTER SETTING sql.stats.automatic_collection.enabled = true;`)
> 	sqlRunner.Exec(t, `SET CLUSTER SETTING sql.stats.automatic_partial_collection.enabled = true;`)
> ```

Since you are manually creating stats below, I don't think you need to enable these settings. It might be better to disable the settings to prevent flakiness in the test.
___
*[`pkg/sql/create_stats_test.go` line 254 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/143283#-OM7m_vuBRhnQJjIdgHX:-OM7m_vuBRhnQJjIdgHY:b-hylxdi) ([raw file](https://github.com/cockroachdb/cockroach/blob/a258eb30f327ab522c5950f2dab6f5565683a373/pkg/sql/create_stats_test.go#L254)):*
> ```Go
> 	sqlRunner.Exec(t, `CREATE TABLE test (id INT PRIMARY KEY, value INT);`)
> 
> 	// Insert more data to increase likelihood of triggering partial stats
> ```

You don't need to increase the likelihood of triggering partial stats, since you are explicitly collecting partial stats below. 

It's not a bad idea to insert data into the table, but I don't think it's strictly necessary given that you're manually collecting stats. I'd either remove this `INSERT` altogether, or update the comment.
___
*[`pkg/sql/create_stats_test.go` line 283 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/143283#-OM7nBKpD7kou_rZ4GXp:-OM7nBKpD7kou_rZ4GXq:b-cebqrw) ([raw file](https://github.com/cockroachdb/cockroach/blob/a258eb30f327ab522c5950f2dab6f5565683a373/pkg/sql/create_stats_test.go#L283)):*
> ```Go
> 		WHERE (job_type = 'AUTO CREATE STATS' OR job_type = 'AUTO CREATE PARTIAL STATS')
> 		AND description LIKE '%test%' 
> 		LIMIT 1`).Scan(&description)
> ```

This will be flaky since it might return a row for `AUTO CREATE STATS` which will have the incorrect description. You should only select job type `AUTO CREATE PARTIAL STATS` in order to avoid failing the check below (ditto above in the `SucceedsSoon` loop).

(Technically this might work if you disable the automatic stats collection cluster setting and only use `ANALYZE`, but it would be misleading.)


<!-- Sent from Reviewable.io -->

veronicavaldez commented:
> Thanks for the contribution, @veronicavaldez! This is definitely on the right track, but there are a few issues with the test that I've called out below. Please also squash your commits.
> 
> > Reviewed 2 of 2 files at r1, 1 of 1 files at r2, all commit messages.
> > _[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/143283#-:-OM7nxiN6dFhCpr3XSlM:b-b0eml0)_ status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @veronicavaldez)
> 
> _[`pkg/sql/create_stats_test.go` line 247 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/143283#-OM7mUYI3nG0AP2GB17C:-OM7mUYI3nG0AP2GB17D:b64roy) ([raw file](https://github.com/cockroachdb/cockroach/blob/a258eb30f327ab522c5950f2dab6f5565683a373/pkg/sql/create_stats_test.go#L247)):_
> 
> > ```go
> > 	sqlRunner := sqlutils.MakeSQLRunner(sqlDB)
> > 
> > 	// Enable both automatic statistics collection and partial stats collection
> > ```
> 
> nit: end comments with a period (here and below).
> 
> _[`pkg/sql/create_stats_test.go` line 249 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/143283#-OM7m9BT5KEFrkQFuLsm:-OM7m9BT5KEFrkQFuLsn:b-q4u0dj) ([raw file](https://github.com/cockroachdb/cockroach/blob/a258eb30f327ab522c5950f2dab6f5565683a373/pkg/sql/create_stats_test.go#L249)):_
> 
> > ```go
> > 	// Enable both automatic statistics collection and partial stats collection
> > 	sqlRunner.Exec(t, `SET CLUSTER SETTING sql.stats.automatic_collection.enabled = true;`)
> > 	sqlRunner.Exec(t, `SET CLUSTER SETTING sql.stats.automatic_partial_collection.enabled = true;`)
> > ```
> 
> Since you are manually creating stats below, I don't think you need to enable these settings. It might be better to disable the settings to prevent flakiness in the test.
> 
> _[`pkg/sql/create_stats_test.go` line 254 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/143283#-OM7m_vuBRhnQJjIdgHX:-OM7m_vuBRhnQJjIdgHY:b-hylxdi) ([raw file](https://github.com/cockroachdb/cockroach/blob/a258eb30f327ab522c5950f2dab6f5565683a373/pkg/sql/create_stats_test.go#L254)):_
> 
> > ```go
> > 	sqlRunner.Exec(t, `CREATE TABLE test (id INT PRIMARY KEY, value INT);`)
> > 
> > 	// Insert more data to increase likelihood of triggering partial stats
> > ```
> 
> You don't need to increase the likelihood of triggering partial stats, since you are explicitly collecting partial stats below.
> 
> It's not a bad idea to insert data into the table, but I don't think it's strictly necessary given that you're manually collecting stats. I'd either remove this `INSERT` altogether, or update the comment.
> 
> _[`pkg/sql/create_stats_test.go` line 283 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/143283#-OM7nBKpD7kou_rZ4GXp:-OM7nBKpD7kou_rZ4GXq:b-cebqrw) ([raw file](https://github.com/cockroachdb/cockroach/blob/a258eb30f327ab522c5950f2dab6f5565683a373/pkg/sql/create_stats_test.go#L283)):_
> 
> > ```go
> > 		WHERE (job_type = 'AUTO CREATE STATS' OR job_type = 'AUTO CREATE PARTIAL STATS')
> > 		AND description LIKE '%test%' 
> > 		LIMIT 1`).Scan(&description)
> > ```
> 
> This will be flaky since it might return a row for `AUTO CREATE STATS` which will have the incorrect description. You should only select job type `AUTO CREATE PARTIAL STATS` in order to avoid failing the check below (ditto above in the `SucceedsSoon` loop).
> 
> (Technically this might work if you disable the automatic stats collection cluster setting and only use `ANALYZE`, but it would be misleading.)



> Thanks for the contribution, @veronicavaldez! This is definitely on the right track, but there are a few issues with the test that I've called out below. Please also squash your commits.
> 
> > Reviewed 2 of 2 files at r1, 1 of 1 files at r2, all commit messages.
> > _[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/143283#-:-OM7nxiN6dFhCpr3XSlM:b-b0eml0)_ status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @veronicavaldez)
> 
> _[`pkg/sql/create_stats_test.go` line 247 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/143283#-OM7mUYI3nG0AP2GB17C:-OM7mUYI3nG0AP2GB17D:b64roy) ([raw file](https://github.com/cockroachdb/cockroach/blob/a258eb30f327ab522c5950f2dab6f5565683a373/pkg/sql/create_stats_test.go#L247)):_
> 
> > ```go
> > 	sqlRunner := sqlutils.MakeSQLRunner(sqlDB)
> > 
> > 	// Enable both automatic statistics collection and partial stats collection
> > ```
> 
> nit: end comments with a period (here and below).
> 
> _[`pkg/sql/create_stats_test.go` line 249 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/143283#-OM7m9BT5KEFrkQFuLsm:-OM7m9BT5KEFrkQFuLsn:b-q4u0dj) ([raw file](https://github.com/cockroachdb/cockroach/blob/a258eb30f327ab522c5950f2dab6f5565683a373/pkg/sql/create_stats_test.go#L249)):_
> 
> > ```go
> > 	// Enable both automatic statistics collection and partial stats collection
> > 	sqlRunner.Exec(t, `SET CLUSTER SETTING sql.stats.automatic_collection.enabled = true;`)
> > 	sqlRunner.Exec(t, `SET CLUSTER SETTING sql.stats.automatic_partial_collection.enabled = true;`)
> > ```
> 
> Since you are manually creating stats below, I don't think you need to enable these settings. It might be better to disable the settings to prevent flakiness in the test.
> 
> _[`pkg/sql/create_stats_test.go` line 254 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/143283#-OM7m_vuBRhnQJjIdgHX:-OM7m_vuBRhnQJjIdgHY:b-hylxdi) ([raw file](https://github.com/cockroachdb/cockroach/blob/a258eb30f327ab522c5950f2dab6f5565683a373/pkg/sql/create_stats_test.go#L254)):_
> 
> > ```go
> > 	sqlRunner.Exec(t, `CREATE TABLE test (id INT PRIMARY KEY, value INT);`)
> > 
> > 	// Insert more data to increase likelihood of triggering partial stats
> > ```
> 
> You don't need to increase the likelihood of triggering partial stats, since you are explicitly collecting partial stats below.
> 
> It's not a bad idea to insert data into the table, but I don't think it's strictly necessary given that you're manually collecting stats. I'd either remove this `INSERT` altogether, or update the comment.
> 
> _[`pkg/sql/create_stats_test.go` line 283 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/143283#-OM7nBKpD7kou_rZ4GXp:-OM7nBKpD7kou_rZ4GXq:b-cebqrw) ([raw file](https://github.com/cockroachdb/cockroach/blob/a258eb30f327ab522c5950f2dab6f5565683a373/pkg/sql/create_stats_test.go#L283)):_
> 
> > ```go
> > 		WHERE (job_type = 'AUTO CREATE STATS' OR job_type = 'AUTO CREATE PARTIAL STATS')
> > 		AND description LIKE '%test%' 
> > 		LIMIT 1`).Scan(&description)
> > ```
> 
> This will be flaky since it might return a row for `AUTO CREATE STATS` which will have the incorrect description. You should only select job type `AUTO CREATE PARTIAL STATS` in order to avoid failing the check below (ditto above in the `SucceedsSoon` loop).
> 
> (Technically this might work if you disable the automatic stats collection cluster setting and only use `ANALYZE`, but it would be misleading.)

Done! I incorporated your suggestions and squashed all commits. Please let me know if there’s anything else I can adjust. Thanks so much for the detailed review @rytaft!

rytaft reviewed (commented):
> Reviewed 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/143283)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @veronicavaldez)

___
*[`pkg/sql/create_stats_test.go` line 283 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/143283#-OM7nBKpD7kou_rZ4GXp:-OMhxlO_E1Zlk_M0DUP6:b-1wyc86) ([raw file](https://github.com/cockroachdb/cockroach/blob/a258eb30f327ab522c5950f2dab6f5565683a373/pkg/sql/create_stats_test.go#L283)):*
<details><summary><i>Previously, veronicavaldez wrote…</i></summary><blockquote>

Done! I incorporated your suggestions and squashed all commits. Please let me know if there’s anything else I can adjust. Thanks so much for the detailed review @rytaft! 
</blockquote></details>

Thanks for the update. Just one remaining issue.
___
*[`pkg/sql/create_stats_test.go` line 258 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/143283#-OMhugYn972iqf-TiV4g:-OMhugYn972iqf-TiV4h:b-mfazth) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b096b017f067f38b6ce2ccac9f4fbf9ce3ad3bc/pkg/sql/create_stats_test.go#L258)):*
> ```Go
> 
> 	// Explicitly create partial stats to ensure we have a job to check.
> 	sqlRunner.Exec(t, `CREATE STATISTICS __auto_partial__ FROM test`)
> ```

This is misleading -- although the test will technically pass (since you're using the name `__auto_partial__`), you aren't actually creating partial stats here anymore. You should first create full stats (either using `ANALYZE test;` as you did in the previous version, or you can fake a full auto stats collection by using `CREATE STATISTICS __auto__ FROM test;`). And then create partial stats using `CREATE STATISTICS __auto_partial__ FROM test USING EXTREMES;`.


<!-- Sent from Reviewable.io -->

veronicavaldez commented:
> > Reviewed 1 of 1 files at r3, all commit messages.
> > _[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/143283)_ status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @veronicavaldez)
> 
> _[`pkg/sql/create_stats_test.go` line 283 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/143283#-OM7nBKpD7kou_rZ4GXp:-OMhxlO_E1Zlk_M0DUP6:b-1wyc86) ([raw file](https://github.com/cockroachdb/cockroach/blob/a258eb30f327ab522c5950f2dab6f5565683a373/pkg/sql/create_stats_test.go#L283)):_
> 
> Previously, veronicavaldez wrote…
> Thanks for the update. Just one remaining issue.
> 
> _[`pkg/sql/create_stats_test.go` line 258 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/143283#-OMhugYn972iqf-TiV4g:-OMhugYn972iqf-TiV4h:b-mfazth) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b096b017f067f38b6ce2ccac9f4fbf9ce3ad3bc/pkg/sql/create_stats_test.go#L258)):_
> 
> > ```go
> > 	// Explicitly create partial stats to ensure we have a job to check.
> > 	sqlRunner.Exec(t, `CREATE STATISTICS __auto_partial__ FROM test`)
> > ```
> 
> This is misleading -- although the test will technically pass (since you're using the name `__auto_partial__`), you aren't actually creating partial stats here anymore. You should first create full stats (either using `ANALYZE test;` as you did in the previous version, or you can fake a full auto stats collection by using `CREATE STATISTICS __auto__ FROM test;`). And then create partial stats using `CREATE STATISTICS __auto_partial__ FROM test USING EXTREMES;`.

Done, thank you again for the feedback!

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> thanks for the contribution!

bors r+

> Reviewed 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/143283#-:-OMxcMMZ-R3jDwMLV5za:bhgc6ko)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @veronicavaldez)



<!-- Sent from Reviewable.io -->
